pkg/solana/chainwriter/chain_writer.go

```diff
@@ -0,0 +1,324 @@
+		if err := json.Unmarshal([]byte(programConfig.IDL), &idl); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal IDL: %w", err)
+		}
+		idlCodec, err := codec.NewIDLInstructionsCodec(idl, binary.LittleEndian())
+		if err != nil {
+			return nil, fmt.Errorf("failed to create codec from IDL: %w", err)
+		}
+		for method, methodConfig := range programConfig.Methods {
+			if methodConfig.InputModifications != nil {
+				modConfig, err := methodConfig.InputModifications.ToModifier(codec.DecoderHooks...)
+				if err != nil {
+					return nil, fmt.Errorf("failed to create input modifications: %w", err)
+				}
+				// add mods to codec
+				idlCodec, err = codec.NewNamedModifierCodec(idlCodec, method, modConfig)
+				if err != nil {
+					return nil, fmt.Errorf("failed to create named codec: %w", err)
+				}
+			}
+		}
+		codecs[program] = idlCodec
+	}
+	return codecs, nil
+}
+
+/*
+GetAddresses resolves account addresses from various `Lookup` configurations to build the required `solana.AccountMeta` list
+for Solana transactions. It handles constant addresses, dynamic lookups, program-derived addresses (PDAs), and lookup tables.
+
+### Parameters:
+- `ctx`: Context for request lifecycle management.
+- `args`: Input arguments used for dynamic lookups.
+- `accounts`: List of `Lookup` configurations specifying how addresses are derived.
+- `derivedTableMap`: Map of pre-loaded lookup table addresses.
+- `debugID`: Debug identifier for tracing errors.
+
+### Return:
+- A slice of `solana.AccountMeta` containing derived addresses and associated metadata.
+
+### Account Types:
+1. **AccountConstant**:
+  - A fixed address, provided in Base58 format, converted into a `solana.PublicKey`.
+  - Example: A pre-defined fee payer or system account.
+
+2. **AccountLookup**:
+  - Dynamically derived from input args using a specified location path (e.g., `user.walletAddress`).
+  - If the lookup table is pre-loaded, the address is fetched from `derivedTableMap`.
+
+3. **PDALookups**:
+  - Generates Program Derived Addresses (PDA) by combining a derived public key with one or more seeds.
+  - Seeds can be `AddressSeeds` (public keys from the input args) or `ValueSeeds` (byte arrays).
+  - Ensures there is only one public key if multiple seeds are provided.
+
+### Error Handling:
+- Errors are wrapped with the `debugID` for easier tracing.
+*/
+// GetAddresses resolves account addresses from various `Lookup` configurations to build the required `solana.AccountMeta` list
+// for Solana transactions.
+func GetAddresses(ctx context.Context, args any, accounts []Lookup, derivedTableMap map[string]map[string][]*solana.AccountMeta, reader client.Reader) ([]*solana.AccountMeta, error) {
+	var addresses []*solana.AccountMeta
+	for _, accountConfig := range accounts {
+		meta, err := accountConfig.Resolve(ctx, args, derivedTableMap, reader)
+		if err != nil {
+			return nil, err
+		}
+		addresses = append(addresses, meta...)
+	}
+	return addresses, nil
+}
+
```

Comment:
I think it's worth adding a comment here, something like: 

```suggestion
// FilterLookupTableAddresses takes a list of accounts and two lookup table maps
// (one for derived tables, one for static tables) and filters out any addresses that are
// not used by the accounts. It returns a map of only those lookup table
// addresses that match entries in `accounts`.
```

Reply:
Added!